File: scala-conductr-bundle-lib/src/main/scala/com/typesafe/conductr/bundlelib/scala/LocationCache.scala
package com.typesafe.conductr.bundlelib.scala

import java.net.{ URI => JavaURI }
import java.util.{ TimerTask, Timer }

import scala.collection.concurrent.TrieMap
import scala.concurrent.ExecutionContext.Implicits
import scala.concurrent.Future
import scala.concurrent.duration.FiniteDuration
import scala.util.Success

/**
 * A structure that describes what we require from a cache specifically for locations in relation to service names.
 * This interface describes an expiring entry cache inspired by Spray Cache in its elements being captured as Futures, thus
 * being able to cope with the thundering herds issue:
 * http://ehcache.org/documentation/2.8/recipes/thunderingherd.
 *
 * Entries that provide a max age duration are scheduled to be removed at that time. The
 * expectation is that this cache is used with such durations. Where there is no duration
 * (this should be rare) then the cache entry is quickly removed after it has been determined.
 * This removal also occurs when the entry cannot be established successfully
 */
trait CacheLike {
  def getOrElseUpdate(serviceName: String)(op: => Future[Option[(JavaURI, Option[FiniteDuration])]]): Future[Option[JavaURI]]
  def remove(serviceName: String): Option[Future[Option[JavaURI]]]
}

object LocationCache {
  def apply() = new LocationCache
}

/**
 * A cache like class as a default implementation.
 */
class LocationCache extends CacheLike {

  private val cache = TrieMap.empty[String, Future[Option[(JavaURI, Option[FiniteDuration])]]]

  val reaperTimer = new Timer("LocationCache-Reaper", true /* daemon thread */ )

  override def getOrElseUpdate(serviceName: String)(op: => Future[Option[(JavaURI, Option[FiniteDuration])]]): Future[Option[JavaURI]] = {
    import Implicits.global
    cache.getOrElseUpdate(serviceName, op)
      .andThen {
        case Success(Some((location, Some(maxAge)))) =>
          reaperTimer.schedule(new TimerTask {
            override def run(): Unit = {
              cache.remove(serviceName)
            }
          }, maxAge.toMillis)

        case _ =>
          cache.remove(serviceName)
      }
      .map(_.map(_._1))

  }

  override def remove(serviceName: String): Option[Future[Option[JavaURI]]] = {
    import Implicits.global
    cache.remove(serviceName)
      .map(_.map(_.map(_._1)))
  }
}

File: java-conductr-bundle-lib/src/main/java/com/typesafe/conductr/bundlelib/java/LocationCache.java
package com.typesafe.conductr.bundlelib.java;

import com.typesafe.conductr.lib.java.Tuple;

import java.net.URI;
import java.time.Duration;
import java.util.NoSuchElementException;
import java.util.Optional;
import java.util.Timer;
import java.util.TimerTask;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;
import java.util.function.Supplier;

/**
 * A cache like class as a default implementation.
 */
public class LocationCache implements CacheLike {

    private final ConcurrentMap<String, CompletionStage<Optional<Tuple<URI, Optional<Duration>>>>> cache = new ConcurrentHashMap<>();

    private final Timer reaperTimer = new Timer();

    @Override
    public CompletionStage<Optional<URI>> getOrElseUpdate(String serviceName, Supplier<CompletionStage<Optional<Tuple<URI, Optional<Duration>>>>> op) {
        return cache
                .computeIfAbsent(serviceName, sn -> op.get())
                .whenCompleteAsync((result, error) -> {
                    /*
                      IMPORTANT: always check the presence of error before result.
                      If error is present, and result is accessed (e.g. calling `isPresent()`), the whole future will
                      fail with an exception.
                     */
                    if (error != null || !result.isPresent())
                        cache.remove(serviceName);
                    else if (result.isPresent()) {
                        if (result.get()._2.isPresent())
                            reaperTimer.schedule(new TimerTask() {
                                @Override
                                public void run() {
                                    cache.remove(serviceName);
                                }
                            }, result.get()._2.get().toMillis());
                        else
                            cache.remove(serviceName);
                    }
                })
                .thenApply(r -> r.map(t -> t._1));
    }

    @Override
    public Optional<CompletionStage<Optional<URI>>> remove(String serviceName) {
        return Optional.ofNullable(
                cache.remove(serviceName)
                    .thenApply(r -> r.map(t -> t._1))
        );
    }
}
